sympy/logic/boolalg.py

```diff
@@ -752,10 +752,11 @@ def _eval_simplify(self, **kwargs):
                 (i.free_symbols, i) for i in eqs]),
                 lambda x: len(x[0]))
             eqs = []
+            nonlineqs = []
             while 1 in sifted:
                 for free, e in sifted.pop(1):
                     x = free.pop()
```

Comment:
`x = next(ordered(free))` would be canonical, but I'm not sure it matters since all single-symbol expressions will eventually be solved. But thinking about this and a recent SO question where recursive resolution of linear expressions is a desirable action, maybe this should be a solve-routine, maybe `extract_known(eqs) -> {xi: vi}, nonlinear_eqs`?

Reply:
OK! Not sure if I get it fully, but say that we have `And(Eq(x**2, x), Eq(x, 1))` or `And(Eq(x**2, 1), Eq(x, 1))`, do we always know that `Eq(x,1)` is processed first? Or `And(Eq(y**2, 1), Eq(x, y), Eq(x, 1))`, isn't there a risk that `Eq(y**2, 1)` is passed on to `eqs` and that `y=1` is never substituted there?

Reply:
Looking more closely. We are dealing with 1 free symbol so there is nothing non-canonical about popping that one symbol from `free` ;-). If we order `(i.free_symbols, i) for i in eqs])` as `(i.free_symbols, i) for i in ordered(eqs)])` then equality `Eq(x, 1)` will come before `Eq(x**2, x)`
```python
>>> list(ordered((Eq(x**2,x),Eq(x,1))))
[Eq(x, 1), Eq(x**2, x)]
```
I am not sure if the equations get made canonical by this step, but if not then
```python
(i.free_symbols, i) for i in [i.canonical for i in ordered(eqs)]])
```
would be best. That will put `Eq(1,x)` as `Eq(x, 1)`.

Reply:
no -- the solving for linear term handles rearrangement. I think this is fine as it is.